#### Aquí encontrarás todo lo que necesitas saber: https://www.kaggle.com/t/ab8726f0cfc84544abbae69a6be88071

## Librerias

In [2]:
import pandas as pd

import urllib.request
from PIL import Image

## Data

In [3]:
### Para que funcione necesitas bajarte los archivos de datos de Kaggle 
train = pd.read_csv("data/train.csv")


## Data exploration

In [5]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 912 entries, 0 to 911
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   laptop_ID         912 non-null    int64  
 1   Company           912 non-null    object 
 2   Product           912 non-null    object 
 3   TypeName          912 non-null    object 
 4   Inches            912 non-null    float64
 5   ScreenResolution  912 non-null    object 
 6   Cpu               912 non-null    object 
 7   Ram               912 non-null    object 
 8   Memory            912 non-null    object 
 9   Gpu               912 non-null    object 
 10  OpSys             912 non-null    object 
 11  Weight            912 non-null    object 
 12  Price_euros       912 non-null    float64
dtypes: float64(2), int64(1), object(10)
memory usage: 92.8+ KB


In [4]:
train.head()

laptop_ID Company            Product   TypeName  Inches   
0       1104    Acer     Aspire ES1-523   Notebook    15.6  \
1        114    Dell             XPS 13  Ultrabook    13.3   
2        172    Acer    Aspire A517-51G   Notebook    17.3   
3        918      HP      Elitebook 820    Netbook    12.5   
4        447  Lenovo  IdeaPad 320-15AST   Notebook    17.3   

                   ScreenResolution                         Cpu  Ram   
0                          1366x768   AMD A8-Series 7410 2.2GHz  4GB  \
1  Quad HD+ / Touchscreen 3200x1800  Intel Core i7 7560U 2.4GHz  8GB   
2       IPS Panel Full HD 1920x1080  Intel Core i5 8250U 1.6GHz  8GB   
3                 Full HD 1920x1080  Intel Core i7 7500U 2.7GHz  8GB   
4                          1600x900   AMD A6-Series 9220 2.5GHz  8GB   

      Memory                           Gpu       OpSys  Weight  Price_euros  
0  500GB HDD                 AMD Radeon R5  Windows 10   2.4kg        387.0  
1  256GB SSD  Intel Iris Plus Graphics 640  Windows 10  1.23kg       1379.0  
2  256GB SSD          Nvidia GeForce MX150  Windows 10     3kg        854.0  
3  512GB SSD         Intel HD Graphics 620  Windows 10  1.26kg       1483.0  
4    1TB HDD                 AMD Radeon R4  Windows 10   2.8kg        519.0

## Data processing

In [44]:
train = pd.read_csv("data/laptops.csv")

-----------------------------------------------------------------------------------------------------------------

## Pasos que si o si debéis realizar para poder participar en la competición:

### 1. Definir X e y

In [42]:
#train = train[['Inches', 'Price_euros']]

In [45]:
X = train.drop(['Price_euros'], axis=1)
y = train['Price_euros'].copy()
X.shape

(912, 17)

In [46]:
y.shape

(912,)

### 2. Dividir X_train, X_test, y_train, y_test

In [47]:
from sklearn.model_selection import train_test_split

In [48]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 42)

In [49]:
X_train

laptop_ID  Company  TypeName  Inches  Ram  OpSys  Weight  Memory1_1   
25          26        2         5    15.6    4      6    2.30       1000  \
84         776        2         2    15.6   16      6    2.72        128   
10        1263        5         2    15.6   16      6    2.50        256   
342        182        4         5    15.6    8      6    1.90        128   
890        321        3         1    13.3   16      6    1.29       1000   
..         ...      ...       ...     ...  ...    ...     ...        ...   
106        394        2         6    15.6    8      6    2.00        256   
270        441        4         5    15.6   16      6    2.50        512   
860        955        5         5    17.3    4      6    2.80       1000   
435        210        2         5    15.6    6      6    2.30       1000   
102        673       10         5    15.6    4      6    2.10        128   

     Memory1_2  Memory2_1  Memory2_2  Resolution  Gpu1  Gpu2  Cpu1  Cpu2  Cpu3  
25           1          0          0        1049     2     3     2     9     9  
84           2       1000          1        8294     3     6     2     9     9  
10           2       1000          1        2073     3     6     2     9     9  
342          2       1000          1        2073     1     8     2     9     9  
890          2          0          0        8294     2     4     2     9     9  
..         ...        ...        ...         ...   ...   ...   ...   ...   ...  
106          2          0          0        2073     3     9     2     9     9  
270          2          0          0        2073     3     9     2     9     9  
860          1          0          0        8294     3     6     2     7     7  
435          1          0          0        1049     2     3     2     9     9  
102          2          0          0        1049     2     3     2     9     9  

[729 rows x 17 columns]

### 3. Asignar el modelo (vacío) a una variable


In [50]:
from sklearn.linear_model import LinearRegression

model = LinearRegression()

In [51]:
model.fit(X_train, y_train)

LinearRegression()

In [52]:
prediction = model.predict(X_test)
prediction

array([1158.28491412,  278.04219786,  881.36022749, 1284.12047324,
       2048.3469203 ,  747.9242513 ,  857.06169006, 1306.84327231,
        888.56831428, 1184.38303388, 1267.20044921,  900.52258131,
        531.23845861, 1226.42752231, 1431.75676851, 1760.62447526,
       1830.23489865, 1309.67942117, 1137.36409688,  485.67694129,
        686.06210568,  220.18731147, 1052.41157111, 1108.43646519,
        958.89227744,  401.78839289, 1023.83730029, 1698.51445504,
       1042.53374957,  284.11207279, 1089.87035418, 1201.99257899,
        704.9698987 , 1315.21469463, 1312.2619886 ,  288.89395983,
        996.81767254,  895.9597329 , 1206.05648689,  789.73239545,
       1189.51254083, 1434.41808593, 1042.77775739, 1789.33402755,
       1152.14069888,  551.6331025 , 1404.52274879,  657.26859701,
       1183.89448588,  458.68658599, 1981.88406004,  801.09512582,
        141.09212359, 1966.33142211,  795.57672221, 1039.72758318,
       1908.37232515, 1136.7507081 , 1773.08654981, 1861.54366

### 4. Sacar métricas, valorar el modelo 

Recuerda que en la competición se va a evaluar con la métrica de MAE.

In [53]:
from sklearn.metrics import mean_squared_error
import numpy as np

print('RMSE :', np.sqrt(mean_squared_error(y_test, prediction)))

RMSE : 449.02179642441564


-----------------------------------------------------------------

## Una vez listo el modelo, toca predecir con el dataset de predicción 

Definición de **modelo que está listo**. 

Tras hacer suficientes pruebas, analizar los datos, hacer feature engineering, probar diferentes modelos con diferentes parámetros, es con este con el que observo mejores métricas y menos overfitting. ¡Cuidado con el overfitting aquí! Si vuestro modelo aprende muy bien de estos datos pero hay overfitting cuando le pasemos los datos desconocidos de `test.csv` nos arriesgamos a que digamos, no salga lo esperado.

### 1. Entrena dicho modelo con TODOS tus datos de train, esto es con `train.csv` al completo.


**CON LAS TRANSFORMACIONES QUE LE HAYAS REALIZADO A `X` INCLUÍDAS.**


Véase:
- Estandarización/Normalización
- Eliminación de Outliers
- Eliminación de columnas
- Creación de columnas nuevas
- Gestión de valores nulos
- Y un largo etcétera de técnicas que como Data Scientist hayas considerado las mejores para tu dataset.

### 2. Carga los datos de `test.csv` para predecir.

In [60]:
X_pred = pd.read_csv("data/test.csv", index_col = 0)
X_pred.head()

Company          Product            TypeName  Inches   
laptop_ID                                                          
750         Toshiba  Tecra X40-D-10G            Notebook    14.0  \
726        Mediacom    Smartbook 142            Notebook    14.0   
633        Mediacom    SmartBook 140            Notebook    14.0   
363            Dell    Inspiron 7577              Gaming    15.6   
319            Asus    VivoBook Flip  2 in 1 Convertible    11.6   

                                    ScreenResolution   
laptop_ID                                              
750        IPS Panel Full HD / Touchscreen 1920x1080  \
726                      IPS Panel Full HD 1920x1200   
633                                Full HD 1920x1080   
363                                Full HD 1920x1080   
319                             Touchscreen 1366x768   

                                            Cpu  Ram              Memory   
laptop_ID                                                                  
750                  Intel Core i5 7200U 2.5GHz  8GB           256GB SSD  \
726                    Intel Atom Z8350 1.92GHz  4GB            32GB HDD   
633                 Intel Atom x5-Z8350 1.44GHz  2GB  32GB Flash Storage   
363                 Intel Core i5 7300HQ 2.5GHz  8GB             1TB HDD   
319        Intel Celeron Dual Core N3350 1.1GHz  2GB  32GB Flash Storage   

                               Gpu       OpSys  Weight  
laptop_ID                                               
750          Intel HD Graphics 620  Windows 10  1.25kg  
726          Intel HD Graphics 400  Windows 10   1.4kg  
633              Intel HD Graphics  Windows 10   1.4kg  
363        Nvidia GeForce GTX 1050  Windows 10  2.65kg  
319          Intel HD Graphics 500  Windows 10   1.1kg

In [61]:
X_pred.shape

(391, 11)

**¿Por qué da error?** 

In [62]:
predictions_submit = model.predict(X_pred)
predictions_submit

ValueError: The feature names should match those that were passed during fit.
Feature names unseen at fit time:
- Cpu
- Gpu
- Memory
- Product
- ScreenResolution
Feature names seen at fit time, yet now missing:
- Cpu1
- Cpu2
- Cpu3
- Gpu1
- Gpu2
- ...


IMPORTANTE: APLICAR LO MISMO A ESTOS DATOS QUE HAYÁIS APLICADO A LOS DATOS DE ENTRENAMIENTO

- SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 4 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS
- SI AL ARRAY CON EL QUE HICISTEIS `.fit()` LO NORMALIZASTEIS, PARA `.predict()` DEBÉIS NORMALIZARLO
- TODO IGUAL SALVO BORRAR FILAS, EL NÚMERO DE ROWS SE DEBE MANTENER EN ESTE SET, PUES LA PREDICCIÓN DEBE TENER 391 FILAS, SI O SI

**Entonces, si al cargar los datos de train usé `index_col=0` para que utilizara la primera columna del conjunto de datos como índice, ¿tendré que hacerlo también para el conjunto `test.csv`?**

In [63]:
X_pred['Weight'] = X_pred['Weight'].replace('kg', '', regex=True)

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

In [19]:
X_pred = pd.read_csv("data/test.csv", index_col=0)
X_pred.head()

Company          Product            TypeName  Inches   
laptop_ID                                                          
750         Toshiba  Tecra X40-D-10G            Notebook    14.0  \
726        Mediacom    Smartbook 142            Notebook    14.0   
633        Mediacom    SmartBook 140            Notebook    14.0   
363            Dell    Inspiron 7577              Gaming    15.6   
319            Asus    VivoBook Flip  2 in 1 Convertible    11.6   

                                    ScreenResolution   
laptop_ID                                              
750        IPS Panel Full HD / Touchscreen 1920x1080  \
726                      IPS Panel Full HD 1920x1200   
633                                Full HD 1920x1080   
363                                Full HD 1920x1080   
319                             Touchscreen 1366x768   

                                            Cpu  Ram              Memory   
laptop_ID                                                                  
750                  Intel Core i5 7200U 2.5GHz  8GB           256GB SSD  \
726                    Intel Atom Z8350 1.92GHz  4GB            32GB HDD   
633                 Intel Atom x5-Z8350 1.44GHz  2GB  32GB Flash Storage   
363                 Intel Core i5 7300HQ 2.5GHz  8GB             1TB HDD   
319        Intel Celeron Dual Core N3350 1.1GHz  2GB  32GB Flash Storage   

                               Gpu       OpSys  Weight  
laptop_ID                                               
750          Intel HD Graphics 620  Windows 10  1.25kg  
726          Intel HD Graphics 400  Windows 10   1.4kg  
633              Intel HD Graphics  Windows 10   1.4kg  
363        Nvidia GeForce GTX 1050  Windows 10  2.65kg  
319          Intel HD Graphics 500  Windows 10   1.1kg

### 3. AHORA puedo hacer la predicción que será lo que subirás a Kaggle. 

**¿Qué es lo que subirás a Kaggle?**

In [27]:
predictions_submit = model.predict(X_pred['Inches'])
predictions_submit

c:\Users\crist\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\base.py:439: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


ValueError: Expected 2D array, got 1D array instead:
array=[14.  14.  14.  15.6 11.6 15.6 17.3 15.6 15.6 15.6 15.6 14.  15.6 15.6
 15.6 15.6 13.3 14.  15.6 15.6 14.  17.3 15.6 15.6 15.6 14.  15.6 17.3
 15.6 15.6 15.6 15.6 15.6 15.6 15.6 13.3 15.6 12.5 14.  13.3 17.3 15.6
 13.3 13.3 15.6 15.6 15.6 17.3 17.3 17.3 15.6 15.6 15.6 15.6 15.6 15.6
 15.6 11.6 13.3 13.3 15.6 15.6 15.6 15.6 15.6 17.3 15.6 17.3 15.6 15.6
 15.6 13.3 17.3 13.3 15.6 15.6 15.6 15.4 15.6 15.6 15.6 15.6 14.  17.3
 17.3 15.6 13.3 13.3 15.6 17.3 13.3 15.6 15.6 13.3 15.6 14.  15.6 15.6
 15.6 15.6 15.6 15.6 15.6 15.6 17.3 15.6 15.6 15.6 15.6 15.6 14.  10.1
 12.5 17.3 13.3 13.3 13.9 17.3 14.  13.3 15.6 14.  17.3 13.3 15.6 15.6
 15.6 12.3 12.5 14.  15.6 15.6 15.6 17.3 10.1 13.3 14.  17.3 15.  14.
 14.  15.6 15.6 15.6 15.6 15.6 17.3 15.6 15.6 15.6 15.6 15.6 13.3 14.
 15.6 15.6 13.9 15.6 11.3 15.6 17.3 15.6 15.6 15.6 15.6 17.3 13.3 15.6
 11.6 15.6 15.6 15.6 15.6 15.6 15.6 15.6 17.3 13.3 12.5 11.6 15.6 15.6
 15.6 15.6 17.3 15.6 13.3 17.3 11.6 15.6 15.6 15.6 13.3 15.6 15.6 17.3
 13.3 15.6 17.3 15.6 12.5 15.6 13.3 15.6 17.3 13.3 13.3 13.3 15.6 14.
 13.3 15.6 15.6 14.  15.6 15.6 15.6 17.3 15.6 15.6 15.6 15.6 14.  15.6
 17.3 14.  12.5 15.6 13.3 13.3 17.3 15.6 15.6 17.3 15.6 13.3 15.6 15.6
 15.6 15.6 14.  11.6 13.  17.3 15.6 14.  12.3 14.  14.  14.  15.6 15.6
 14.  15.6 11.6 15.6 13.3 13.3 12.5 15.6 14.  14.  14.  15.6 15.6 13.3
 15.6 15.6 13.3 15.6 15.6 15.6 15.6 15.6 17.3 15.6 17.3 15.6 15.6 17.3
 13.3 13.3 17.3 17.3 15.6 15.6 15.6 17.3 17.3 13.3 15.6 15.6 15.6 15.6
 15.6 17.3 11.6 17.3 13.3 14.  15.6 15.6 15.6 17.3 14.  15.6 14.  15.6
 17.3 13.3 12.5 15.6 15.6 13.3 15.6 13.3 17.3 14.  13.3 15.6 14.  15.6
 17.3 17.3 15.6 17.3 14.  15.6 15.6 15.6 13.3 15.6 17.3 15.6 15.6 15.6
 15.6 15.6 14.  15.6 14.  15.6 15.6 15.6 15.6 15.6 15.6 17.3 15.6 15.
 15.6 14.  15.6 15.6 15.6 14.  14.  17.3 17.3 15.6 13.3 15.6 15.6 17.3
 14.  15.6 15.6 15.6 14.  15.6 11.6 15.6 17.3 15.6 15.6 14.  15.6 14.
 15.6 13.3 11.6 13.3 11.6 15.6 15.6 15.6 15.6 12.5 15.6 15.6 12.5].
Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.

**¡PERO! Para subir a Kaggle la predicción, ésta tendrá que tener una forma específica y no valdrá otra.**

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [20]:
sample = pd.read_csv("data/sample_submission.csv")

In [21]:
sample.head()

laptop_ID  Price_euros
0        750          500
1        726          500
2        633          500
3        363          650
4        319          650

In [22]:
sample.shape

(391, 2)

### 4. Mete tus predicciones en un dataframe. 

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [23]:
submission = pd.DataFrame({"laptop_ID": sample['laptop_ID'], "Price_euros": predictions_submit})

NameError: name 'predictions_submit' is not defined

In [27]:
submission.head()

id  Price_euros
0  1184         2158
1   815          511
2  1244         5290
3  1121          622
4   443         1140

In [28]:
submission.shape

(391, 2)

### 5. Pásale el CHEQUEATOR para comprobar que efectivamente está listo para subir a Kaggle.

In [28]:
def chequeator(df_to_submit):
    """
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.
    """
    if df_to_submit.shape == sample.shape:
        if df_to_submit.columns.all() == sample.columns.all():
            if df_to_submit.laptop_ID.all() == sample.laptop_ID.all():
                print("You're ready to submit!")
                submission.to_csv("submission.csv", index = False) #muy importante el index = False
                urllib.request.urlretrieve("https://i.kym-cdn.com/photos/images/facebook/000/747/556/27a.jpg", "gfg.png")     
                img = Image.open("gfg.png")
                img.show()   
            else:
                print("Check the ids and try again")
        else:
            print("Check the names of the columns and try again")
    else:
        print("Check the number of rows and/or columns and try again")
        print("\nMensaje secreto de Clara: No me puedo creer que después de todo este notebook hayas hecho algún cambio en las filas de `diamonds_test.csv`. Lloro.")


In [29]:
chequeator(submission)

NameError: name 'submission' is not defined